### PR TITLE
[megatron] fix: MTP compatible with latest mcore

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

- Base mcore with https://github.com/Nvidia/Megatron-LM/commit/b57449928aafaaec7a36e8325d83aacf05e91bd4 by updating patches and ensure backward compatible
- Also make mtp rollout support passing additional rollout_engine_kwargs
- Also make request_spec_decode_stats optional

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [X] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/models/mcore/mtp_patch.py [hunk 1]
```diff
@@ -100,6 +106,9 @@ def _megatron_gptmodel_postprocess(
         output_weight = self.shared_embedding_or_output_weight()
 
     if mtp_in_postprocess and labels is not None:
+        mtp_kwargs = dict(extra_block_kwargs or {})
+        if "padding_mask" in signature(self.mtp.forward).parameters:
+            mtp_kwargs["padding_mask"] = padding_mask
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:113]: fixed

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 2]
```diff
@@ -296,9 +297,13 @@ async def launch_server(self, master_address: str = None, master_port: int = Non
                 args["served_model_name"] = served_model_name
 
         if self.config.mtp is not None and self.config.mtp.enable and self.config.mtp.enable_rollout:
+            mtp_engine_kwargs = (
+                self.config.mtp.get("rollout_engine_kwargs", {}).get(self._get_engine_kwargs_key(), {}) or {}
+            )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py]: fixed

File: verl/models/mcore/mtp_patch.py [hunk 3]
```diff
@@ -267,10 +288,19 @@ def patch_mtp_layer_checkpointed_forward(model: torch.nn.Module):
...
+                params = ["forward_func"]
+            if not params or params[0] != "forward_func":
+                continue
             layer._checkpointed_forward_backup = layer._checkpointed_forward
             layer._checkpointed_forward = _patched_checkpointed_forward.__get__(layer, layer.__class__)
-        print(f"Found and patched checkpointed forward for {len(target_layers)} MTP layer(s)")
+            patched_count += 1
+        if patched_count:
+            print(f"Found and patched checkpointed forward for {patched_count} MTP layer(s)")
         return True
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py]: fixed

File: verl/models/mcore/mtp_patch.py [hunk 4]
```diff
@@ -127,20 +138,29 @@ def _megatron_gptmodel_postprocess(
...
+                "config": self.config,
+                "cp_group": cp_group,
+                "packed_seq_params": packed_seq_params,
+                "scale_logits_fn": scale_logits_fn,
+                "input_ids": input_ids,
+            }
+            if "tp_group" in _PROCESS_MTP_LOSS_PARAMS:
+                process_mtp_loss_kwargs["tp_group"] = getattr(self, "tp_group", None) or (
+                    self.pg_collection.tp if hasattr(self, "pg_collection") else None
+                )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:158]: fixed

File: verl/models/mcore/mtp_patch.py [hunk 5]
```diff
@@ -31,8 +32,10 @@
     from megatron.core.transformer.multi_token_prediction import process_mtp_loss as _process_mtp_loss
 
     _HAS_PROCESS_MTP_LOSS = True
+    _PROCESS_MTP_LOSS_PARAMS = set(signature(_process_mtp_loss).parameters)
 except ImportError:
     _HAS_PROCESS_MTP_LOSS = False
+    _PROCESS_MTP_LOSS_PARAMS = set()
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:38]: unneeded

File: verl/models/mcore/mtp_patch.py [hunk 6]
```diff
@@ -100,6 +106,11 @@ def _megatron_gptmodel_postprocess(
         output_weight = self.shared_embedding_or_output_weight()
 
     if mtp_in_postprocess and labels is not None:
+        mtp_kwargs = dict(extra_block_kwargs or {})
+        if not hasattr(self.mtp, "_forward_has_padding_mask"):
+            self.mtp._forward_has_padding_mask = "padding_mask" in signature(self.mtp.forward).parameters
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:111]: over defensive

File: verl/models/mcore/mtp_patch.py [hunk 7]
```diff
@@ -235,6 +254,7 @@ def patch_mtp_layer_get_embeddings(model: torch.nn.Module):
 
     if target_layers:
         for layer in target_layers:
+            layer._get_embeddings_has_padding_mask = "padding_mask" in signature(layer._get_embeddings).parameters
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:260]: over defensive

File: verl/trainer/config/model/hf_model.yaml [hunk 8]
```diff
@@ -95,6 +95,9 @@ mtp:
 
   method: mtp
   num_speculative_tokens: 1
+  rollout_engine_kwargs:
```
Comment (wuxibin89) [inline verl/trainer/config/model/hf_model.yaml]: We should reuse rollout.engine_kwargs.{vllm|sglang|trtllm}:
https://github.com/verl-project/verl/blob/main/verl/trainer/config/rollout/rollout.yaml#L130-L140

**Review by wuxibin89 — CHANGES_REQUESTED** [review]

File: verl/trainer/config/model/hf_model.yaml (hunk 8 above)
Comment (HollowMan6) [reply verl/trainer/config/model/hf_model.yaml]: Now I removed this new field and fix the code so that user specified speculative_config no longer gets overwritten: https://github.com/verl-project/verl/pull/6648/changes/0f06986232d9dd0abf45bb9b8268a42cd4e26600

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 9]
```diff
@@ -587,12 +589,18 @@ async def generate(
 
         # Re-key backend spec-decoding stats to the rollout-common names.
         if self.config.mtp is not None and self.config.mtp.enable and self.config.mtp.enable_rollout:
-            if final_res.metrics is None or final_res.metrics.request_spec_decode_stats is None:
-                raise RuntimeError("vLLM MTP rollout requires request_spec_decode_stats; set disable_log_stats=False.")
-            spec_decode_stats = final_res.metrics.request_spec_decode_stats
-            extra_fields["spec_num_draft_tokens"] = spec_decode_stats.num_draft_tokens
-            extra_fields["spec_num_accepted_tokens"] = spec_decode_stats.num_accepted_tokens
-            extra_fields["spec_num_verify_steps"] = spec_decode_stats.num_verify_steps
+            spec_decode_stats = getattr(final_res.metrics, "request_spec_decode_stats", None)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py:592]: this is over defensive

File: verl/models/mcore/mtp_patch.py [hunk 10]
```diff
@@ -267,11 +287,20 @@ def patch_mtp_layer_checkpointed_forward(model: torch.nn.Module):
...
+        patched_count = 0
         for layer in target_layers:
+            try:
+                params = list(signature(layer._checkpointed_forward).parameters)
+            except (TypeError, ValueError):
+                params = ["forward_func"]
+            if not params or params[0] != "forward_func":
+                continue
             layer._checkpointed_forward_backup = layer._checkpointed_forward
             layer._checkpointed_forward = _patched_checkpointed_forward.__get__(layer, layer.__class__)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:306]: fixed

File: verl/models/mcore/mtp_patch.py [hunk 11]
```diff
@@ -235,6 +254,7 @@ def patch_mtp_layer_get_embeddings(model: torch.nn.Module):
 
     if target_layers:
         for layer in target_layers:
+            layer._get_embeddings_has_padding_mask = "padding_mask" in signature(layer._get_embeddings).parameters
             layer._get_embeddings_backup = layer._get_embeddings
             layer._get_embeddings = _patched_get_embeddings_for_detach.__get__(layer, layer.__class__)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:262]: fixed

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py:592]: this is over defensive

File: verl/models/mcore/mtp_patch.py [hunk 12]
```diff
@@ -234,11 +253,17 @@ def patch_mtp_layer_get_embeddings(model: torch.nn.Module):
                 target_layers.append(layer)
 
     if target_layers:
+        patched_count = 0
         for layer in target_layers:
+            if hasattr(layer, "_get_embeddings_backup"):
+                continue
+            layer._get_embeddings_has_padding_mask = "padding_mask" in signature(layer._get_embeddings).parameters
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/models/mcore/mtp_patch.py:260]: this is over defensive

**Review by wuxibin89 — APPROVED** [review]
